Feat(CORE-941): Add using the library page (#838)
gruntwork-io/docs: _docs-sources/iac/usage/using-the-library.md

go get github.com/stretchr/testify/assert
go get github.com/aws/aws-sdk-go/aws
go mod tidy
```

### Write the test

Next, we’ll write the test. Specify a single test called `TestLambdaCreated` that provisions an AWS Lambda function, confirms it is created, then destroys the Lambda function. We’ll use some of the built-in functionality in `Terratest` to generate random values and set variables that will be passed into Terraform.

```go title=gw_module_guide/test/lambda_test.go
package test

package test

import (
	"os"
	"testing"

	"fmt"

	awsSDK "github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/service/lambda"
	"github.com/gruntwork-io/terratest/modules/aws"
	"github.com/gruntwork-io/terratest/modules/random"
	"github.com/gruntwork-io/terratest/modules/terraform"
	"github.com/stretchr/testify/assert"
)

func TestLambdaCreated(t *testing.T) {
	// Run this test in parallel with all the others
	t.Parallel()

	// Unique ID to namespace resources
	uniqueId := random.UniqueId()
	// Generate a unique name for each Lambda so any tests running in parallel don’t clash
	lambdaName := fmt.Sprintf("test-lambda-%s", uniqueId)

	// Get the cwd so we can point to the lambda handler

> 	path, _ := os.Getwd()

Reviewer: NIT: handle the error here

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Using the Library

> The Terraform modules in the Gruntwork Infrastructure as Code Library allows you to configure the `provider` and `backend` settings to suit their needs. This makes it possible to use Gruntwork modules alongside any existing modules you may have already developed, with minimal duplication of configuration.

Reviewer: ```suggestion
The Terraform modules in the Gruntwork Infrastructure as Code Library allow you to configure the `provider` and `backend` settings to suit your needs. This makes it possible to use Gruntwork modules alongside any existing modules you may have already developed, with minimal duplication of configuration.
```

## Prerequisites

- An AWS account with permissions to create the necessary resources
- An [AWS Identity and Access Management](https://aws.amazon.com/iam/) (IAM) user or role with permissions to create AWS IAM roles, Lambda functions, and Cloudwatch Log Groups
- [AWS Command Line Interface](https://aws.amazon.com/cli/) (AWS CLI) installed on your local machine
- [Terraform](https://www.terraform.io) installed on your local machine
- (Optional — only required for testing) [Go](https://go.dev) installed on your local machine

## Create a module

In this section you’ll create a module that provisions an AWS Lambda Function using the [`terraform-aws-lambda`](../../reference/modules/terraform-aws-lambda/lambda/) Gruntwork module. This module automatically creates the AWS IAM role and Cloudwatch Log Group for the Lambda function. For more information about the configuration of the module, refer to the [Library Reference](../../reference/modules/terraform-aws-lambda/lambda/#reference).

### Create the basic file structure

> First, create the basic file structure required that will contain the module reference. In this guide, you will create a module named `serverless-api` which contains the reference to the `terraform-aws-lambda` module. This approach is taken to allow you to define a module once, then use it in many environments and regions, while keeping the amount of copy-and-paste work to a minimum.

Reviewer: ```suggestion
First, create the basic file structure that will contain the module reference. In this guide, you will create a module named `serverless-api` which contains the reference to the `terraform-aws-lambda` module. This approach is taken to allow you to define a module once, then use it in many environments and regions, while keeping the amount of copy-and-paste work to a minimum.
```

Reviewer: NIT: passive voice in "This approach is taken..."

touch gw_module_guide/example/<YOUR_REGION>/main.tf

mkdir -p gw_module_guide/example/<YOUR_REGION>/src
touch gw_module_guide/example/<YOUR_REGION>/src/main.py
```

</TabItem>
<TabItem value="Terragrunt" label="Terragrunt" default>

For Terragrunt, we recommend that all re-usable infrastructure modules live in a directory called `_envcommon`. Create two paths — one which will contain the reference to the `terraform-aws-lambda` and one which will contain a reference to the local module.

```bash
mkdir -p gw_module_guide/_envcommon/serverless-api
touch gw_module_guide/terragrunt.hcl
touch gw_module_guide/_envcommon/serverless-api/lambda.hcl

mkdir -p gw_module_guide/example/<YOUR REGION>/example/serverless-api
touch gw_module_guide/example/<YOUR REGION>/example/serverless-api/terragrunt.hcl

mkdir -p gw_module_guide/example/<YOUR REGION>/example/serverless-api/src
touch gw_module_guide/example/<YOUR REGION>/example/serverless-api/src/main.py
```

</TabItem>
</Tabs>

### Create the reference to the Gruntwork module

Next, create the reference to the `terraform-aws-lambda` module...

<Tabs groupId="tool-choice">
<TabItem value="Terraform" label="Terraform" default>

> Define a module block that uses the git url `terraform-aws-lambda` module for the `source` attribute.

Reviewer: ```suggestion
Define a module block that uses the git url of the `terraform-aws-lambda` module for the `source` attribute.
```

Author: I did some re-ordering of the wording here, moving the common paragraph up above into the section header. LMK what you think.

```hcl title=gw_module_guide/serverless-api/lambda/main.tf
module "lambda" {
  source = "git::git@github.com:gruntwork-io/terraform-aws-lambda.git//modules/lambda?ref=v0.21.9"

  name        = var.name
  runtime     = var.runtime
  source_path = var.source_path
  handler     = var.handler
  run_in_vpc  = false
  timeout     = 30
  memory_size = 128
}
```

Next, add the variables to the `variables.tf` file.

```hcl title=gw_module_guide/serverless-api/lambda/variables.tf
variable "name" {
  type        = string
  description = "Name that will be used for the AWS Lambda function"
}

variable "runtime" {
  type        = string
  description = "The runtime of the Lambda. Options include go, python, ruby, etc."
}

variable "source_path" {
  type        = string
  description = "The path to the directory containing the source to be deployed to lambda"
}

variable "handler" {
  type        = string
  description = "The name of the handler function that will be called as the entrypoint of the lambda"
}

```

</TabItem>
<TabItem value="Terragrunt" label="Terragrunt" default>

> Define the module in `_envcommon` that references the `terraform-aws-lambda`. For `terragrunt`, the source is defined in the `source` attribute of the `terraform` block.

Reviewer: ```suggestion
Define the module in `_envcommon` that references the `terraform-aws-lambda` module. For `terragrunt`, the source is defined in the `source` attribute of the `terraform` block.
```

Reviewer: Feels like an overuse of the word module, but without it, it also sounds strange to me.

Author: Good catch - this is what i meant it to say :lolsob:

> One of the benefits of referencing modules this way is that ability to set defaults for your organization. As an example — the `terraform-aws-lambda` exposes many variables but in the module block below, we are hard coding the value `run_in_vpc` to be `false`. This will ensure that anyone consuming this module will only create AWS Lambda functions that are not in a VPC. For a full list of configuration options for this module, refer to the [Library Reference](../../reference/modules/terraform-aws-lambda/lambda/#reference).

Reviewer: ```suggestion
One of the benefits of referencing modules this way is the ability to set defaults for your organization. As an example — the `terraform-aws-lambda` exposes many variables but in the module block below, we are hard coding the value `run_in_vpc` to be `false`. This will ensure that anyone consuming this module will only create AWS Lambda functions that are not in a VPC. For a full list of configuration options for this module, refer to the [Library Reference](../../reference/modules/terraform-aws-lambda/lambda/#reference).
```

<Tabs groupId="tool-choice">
<TabItem value="Terraform" label="Terraform" default>

Create a module block that uses the path to the local module as the source attribute, supplying values for the required attributes of the module.

```hcl title=gw_module_guide/example/<YOUR_REGION>/main.tf
module "my_lambda" {
  source = "../../serverless-api/lambda"

  name        = "gruntwork-lambda-module-guide"
  runtime     = "python3.9"
  source_path = "${path.module}/src"
  handler     = "main.lambda_handler"
}
```

Next, copy the following python code which will be used as the entrypoint of the AWS Lambda function.

```python title=gw_module_guide/example/<YOUR_REGION>/src/main.py
def lambda_handler(event, context):
    return "Hello from Gruntwork!"
```

</TabItem>
<TabItem value="Terragrunt" label="Terragrunt" default>

Create a module block that uses the path to the local module as the source attribute, supplying values for the required attributes of the module.

> One of the benefits of this approach is that you can increment the version of a module for a specific environment and region in a granular fashion. For example, when we shipped a version v0.22.0 of the `terraform-aws-lambda` module, you could update just the `example` environment in the us-west-2 AWS region to ensure the upgrade goes as expected, then roll out to other environments or regions.

Reviewer: ```suggestion
One of the benefits of this approach is that you can increment the version of a module for a specific environment and region in a granular fashion. For example, if we were to ship a version v0.22.0 of the `terraform-aws-lambda` module, you could update just the `example` environment in the us-west-2 AWS region to ensure the upgrade goes as expected, then roll out to other environments or regions.
```

	// Create a lambda client so we can retrieve the function
	lambdaClient := aws.NewLambdaClient(t, "us-west-2")
	function, _ := lambdaClient.GetFunction(&lambda.GetFunctionInput{
		FunctionName: &lambdaName,
	})

	// Assert the function name is equal to what we set
	assert.Equal(t, lambdaName, awsSDK.StringValue(function.Configuration.FunctionName))
}
```

In this test, we first generate data so that the test run creates resources with unique names. Next, we create the Terraform `options`, which indicates the folder in which the Terraform module we want to test is located and sets the values that will be passed in for variables. Then, we set up a `terraform destroy` operation, which will always run regardless of the test status. Then, we run `terraform init` and `terraform apply` to create the resources. Finally, we validate that the name of the AWS Lambda function that was created matches the expected name.

### Run the test

Finally, run the test you wrote. From the `test` directory, run the following command:
```bash
go test -v
```

You should expect to see `--- PASS: TestLambdaCreated` in the final log lines of the output from the test.

## What’s next

> Now that you’ve used a Gruntwork module to provision resources, consider how you would need to expand this usage make the Lambda function available via a URL using an [AWS API Gateway HTTP API](../../reference/modules/terraform-aws-lambda/api-gateway-proxy/). We refer to combining multiple modules into a single deliverable module as a [service](../overview/services.md).

Reviewer: ```suggestion
Now that you’ve used a Gruntwork module to provision resources, consider how you would need to expand this usage to make the Lambda function available via a URL using an [AWS API Gateway HTTP API](../../reference/modules/terraform-aws-lambda/api-gateway-proxy/). We refer to combining multiple modules into a single deliverable module as a [service](../overview/services.md).
```

> <!-- Leaving this commented out since these two docs pages are coming soon -->

Reviewer: ```suggestion
```

Author: Removed in https://github.com/gruntwork-io/docs/pull/838#:~:text=Verified-,5d6ac59,-Merge%20state
